controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_types.go

```diff
@@ -234,6 +239,20 @@ type RemediationStrategy struct {
 	MinHealthyPeriod *metav1.Duration `json:"minHealthyPeriod,omitempty"`
 }
 
+// MachineNamingStrategy allows changing the naming pattern used when creating Machines.
+// InfraMachines & KubeadmConfigs will use the same name as the corresponding Machines.
+type MachineNamingStrategy struct {
+	// Template defines the template to use for generating the names of the Machine objects.
+	// If not defined, it will fallback to `{{ .kubeadmcontrolplane.name }}-{{ .random }}`.
+	// If the templated string exceeds 63 characters, it will be trimmed to 58 characters and will
+	// get concatenated with a random suffix of length 5.
+	// The templating mechanism provides the following arguments:
+	// * `.kubeadmcontrolplane.name`: The name of the KubeadmControlPlane object.
+	// * `.random`: A random alphanumeric string, without vowels, of length 5.
+	// +optional
+	Template string `json:"template,omitempty"`
```

Comment:
You should impose a maximal length on this string, you know that the rendered value must not exceed 63 chars, so it needs to be at least that in length, assuming a worst case, the value of any substituted variable could be 1 character right?

So it needs to be 62 plus the length in the template of representing the longest variable. Currently that is `{{ .kubeadmcontrolplane.name }}` which is 31 characters, so, a limit of 93 would be suitable presently

Reply:
I think we can do this slightly better and easier. We already have similar validation for the other namingStrategies we have: https://github.com/kubernetes-sigs/cluster-api/blob/8752f4f91301b970c8f50f2bb9c39f5a50caea5d/internal/webhooks/clusterclass.go#L437

So we can verify:
* that the template actually renders
* that the rendered output is a valid name
* use the minimal length for our parameters and then check if the generated output is too long
  * Not sure if we actually ever hit that case, because we have that safeguard that we just cut if rendered string gets too long. So I think we don't need this validation?

Reply:
Sorry I am a bit confused, do we need restrictions on minimum chars too for the generated name?

Reply:
@JoelSpeed wdyt based on my comment above?

Reply:
I think that sounds reasonable yes! So that would be webhook based validation, you should still also include an API side maximum length though (since we can calculate a reasonable maximum)

Having maximum lengths on all fields is good practice and helps with the runtime cost analysis that's built into the CEL libraries, should we ever get to using those.

You can add length limits when the API is introduced, restricting them later is harder

Reply:
Ah yeah, good point. I forgot the CEL part

Reply:
The validation webhook that is in clusterclass I believe is checking all the things needed. I have added the same here, let me know if anything needs to be updated.

Reply:
I can see that the webhook is checking the template, but, I think we do still want to have a length limit as a marker on the API field here, for completeness

Reply:
I have added the marker for max length please check.

Reply:
Question: shouldn't it be more than 93?

Example:

* `{{ .cluster.name }}` (len 19) results in `a`
* `{{ .kubeadmcontrolplane.name }}` (len 31) results in `b`
* `{{ .random }}` (len 14) results in `ccccc`

and a template uses all of these three, e.g.:

* `{{ .cluster.name }}{{ .kubeadmcontrolplane.name }}{{ .random }}` (len 63) results in `abccccc` with length 7
* So the template could use up to 62-7=55 more characters.
* So the MaxLength should be `55+63=118`?

We could say it should be either
* `{{ .cluster.name }}{{ .random }}` (len 32) results in `accccc`
* `{{ .kubeadmcontrolplane.name }}{{ .random }}` (len 45) results in `bccccc`
* In this case the maximum should be `62-6 + 45 = 101`

I know its very theoretical :-)

Reply:
I am leaning toward first option(118 one),  
I think it should be 119 not 118, the generated string should not exceed `63 chars`, so it should be `63-7=56` and `55+63=119`

Reply:
Honestly, Folks can write a lot more than this in Go templates (e.g. if else etc.) Is it really necessary to optimize this on that level? Maybe let's just pick 256 and we're done with it

Reply:
I was thinking something similar tbh, I'd be ok with 256, it gives some bound that we are relatively confident isn't going to cause an issue, if a user complains, we can always increase it later

Reply:
I have changed it to 256.